src/extension/ai/generate.ts

```diff
@@ -0,0 +1,85 @@
+import * as vscode from 'vscode'
+// import {FoyleClient, getTraceID} from './client';
+// import * as converters from './converters';
+// generateCompletion generates a completion by calling the foyle backend and then adds
+// the returned blocks to the window
+import {
+  GenerateCellsRequest,
+  GenerateCellsResponse,
+} from '@buf/stateful_runme.community_timostamm-protobuf-ts/runme/ai/v1alpha1/ai_pb'
+import {
+  Notebook,
+  Cell,
+} from '@buf/stateful_runme.community_timostamm-protobuf-ts/runme/parser/v1/parser_pb'
+
+import { initAIServiceClient } from './client'
+import * as converters from './converters'
+
+export async function generateCompletion() {
+  const editor = vscode.window.activeNotebookEditor
+
+  if (!editor) {
+    return
+  }
+
+  if (editor?.selection.isEmpty) {
+    return
+  }
+
+  // We subtract 1 because end is non-inclusive
+  const lastSelectedCell = editor?.selection.end - 1
+  var lastActiveCell = editor?.notebook.cellAt(lastSelectedCell)
+  console.log('Getting cells')
+  let cells = editor?.notebook.getCells(new vscode.NotebookRange(0, editor?.notebook.cellCount))
+  console.log('Got cells')
+
+  const req = GenerateCellsRequest.create()
+  req.notebook = Notebook.create()
+  req.notebook.cells = []
+  for (let cell of cells) {
+    let cellPb = converters.cellDataToProto(converters.cellToCellData(cell))
+    req.notebook.cells.push(cellPb)
+  }
+
+  let client = initAIServiceClient()
+
+  client
+    .generateCells(req)
+    .then((finished) => {
+      let response = finished.response
+      // TODO(jeremy): We should have the server add the traceId to the response and then we should
```

Comment:
There's a logger abstraction that makes logging human-friendly with most of the bells and whistles, e.g. here https://github.com/stateful/vscode-runme/blob/main/src/extension/server/runmeServer.ts#L38

Reply:
There's only a simple `runme.server.enableLogger` bool switch that will enable debug logging to `Output->Runme`.